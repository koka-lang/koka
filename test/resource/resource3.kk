// Test resource handler

effect heap<s::S> { }

fun heap( action : forall<s> () -> <heap<s>|e> a ) : e a {
  handle<heap<_>>(action) { }
}

effect resource ref<s::S,a::V> in heap<s> {
  fun get() : a
  fun set( value : a ) : ()
}

val new-ref = handler resource(s) {
  get() -> resume(s,s)
  set(x) -> resume((),x)
}

fun main() {
  using heap
  use r1 = new-ref(40)
  use r2 : ref<_,string> = new-ref("hi")
  println(r1.get + r2.get.count)
}