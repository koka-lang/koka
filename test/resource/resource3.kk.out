[01;30moperator branch (new-ref) get: resume tail
[00;39;49m[00m[01;30moperator branch (new-ref) set: resume tail
[00;39;49m[00m42
.Con-Ref : forall<s,a> (resource<heap<s>>) -> ref<s,a>
.Op-get : forall<s,a> .op-get<s,a>
.Op-set : forall<s,a> (value : a) -> .op-set<s,a>
.Ops-get : forall<s,a,b> (get : .op-get<s,a>) -> .ops-ref<s,a,a>
.Ops-set : forall<s,a,b> (set : .op-set<s,a>) -> .ops-ref<s,a,()>
.create-ref : forall<s,a> (int) -> ref<s,a>
.resource-ref: forall<s,a> (ref<s,a>) -> int
.tag-heap : string
.tag-ref : string
get : forall<s,a> (ref<s,a>) -> <div,heap<s>,exn> a
heap : forall<a,e> (action : forall<s> () -> <heap<s>|e> a) -> e a
inject-resource: forall<a,s,b,e> (resource : ref<s,a>, action : () -> <heap<s>|e> b) -> <heap<s>|e> b
main : () -> <pure,console> ()
new-ref : forall<a,s,b,e> (s : a, action : (resource : ref<s,a>) -> <heap<s>,div|e> b) -> <heap<s>,div|e> b
set : forall<s,a> (ref<s,a>, value : a) -> <div,heap<s>,exn> ()